### DISCUSSION TITLE
RTC prescaler | Attiny816

### DISCUSSION THREAD

**Discussion by ASDosjani** [SpenceKonde/megaTinyCore#1113] [Q&A] [answered]
Hi everyone, I'm trying to make the RTC prescaler work but even though I set up the RTC.CTRLA register as datasheet says, nothing changes. I use the example which is mentioned in:
https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md

If anyone has a solution or a good example how to make the attiny sleep more than 32s(the max without prescaler) please tell. I already write a function which counts wake ups and sleep the MCU back if the desired sleep time is not reached, but it's still waking up the MCU to sleep it back.
Thanks in advance for your help.

@SpenceKonde you might know this.

**Comment by hmeijdam** [comment, marked as answer]
I never use the sleep function, but you can prescale the RTC to trigger an interrupt for example once in a couple of hours

The PIT can only do once every 32 seconds max, but the  RTC_CNT interrupt can be set for hours. In below example I set the RTC to use the 1KHz clock signal and then count to 1000. Without prescaling that takes one second. In the example I prescale with 64, so it takes 64 seconds between a RTC_CNT interrupt to occur. Visualizeing by toggling a led on PB0.

If you set it to `RTC_CTRLA |= RTC_PRESCALER_DIV32768_gc | RTC_RTCEN_bm; // turn on RTC ansd set RTC prescaler factor` It is going to take 32768 seconds, which is a little over 9 hours.

If that's still not enough, you increase the counter to count to 5000, to get 45 hours, etc...

```
void setup() {
  //configure output pins
  VPORTA_DIR |= PIN6_bm;   // set pin PA6 as output
  VPORTB_DIR |= PIN0_bm;   // set pin PB0 as output

  //Configure clock source used by both PIT and RTC
  //RTC_CLKSEL = RTC_CLKSEL_INT32K_gc; /* Internal 32kHz OSC */
  RTC_CLKSEL = RTC_CLKSEL_INT1K_gc; /* Internal 1kHz OSC */
  //RTC_CLKSEL = RTC_CLKSEL_EXTCLK_gc;  // Use EXTCLK pin as clock source (connect to CLKOUT)

  //Configure RTC
  while (RTC_STATUS & RTC_PERBUSY_bm) {} // wait until RTC_PER is available for writing
  RTC_PER = 1000; // write period (16bit) to period (1000 is one sec if 1KHz clock used and no prescaler factor)
  //  RTC_PER = 50000; // write period (16bit) to period
  RTC_INTCTRL |= RTC_OVF_bm; // activate overflow interrupts
  RTC_CTRLA |= RTC_PRESCALER_DIV64_gc | RTC_RTCEN_bm; // turn on RTC ansd set RTC prescaler factor
  // possible RTC prescaler factors are DIV1, DIV2, DIV4, DIV8, DIV16, ........DIV16384, DIV32768

  //Configure PIT
  RTC_PITINTCTRL |= RTC_PI_bm;    // activate PIT Interrupt enable
  RTC_PITCTRLA   |= RTC_PERIOD_CYC32768_gc | RTC_PITEN_bm;// set PIT period and enable
}

void loop() {} // nothing to do here

ISR(RTC_PIT_vect) {             // RTC Interrupt
  RTC_PITINTFLAGS = RTC_PI_bm;  // clear  Interrupt Flag
  PORTA_OUTTGL = PIN6_bm;       // toggle Pin/A6
}

ISR(RTC_CNT_vect) {             // RTC Interrupt
  RTC_INTFLAGS = RTC_OVF_bm;  // clear  Interrupt Flag
  PORTB_OUTTGL = PIN0_bm;       // toggle Pin/B0
}


//The Realtimer-Counter (RTC) have a own 32,768KHz Cristal.
// 32.768KHz / CYCprescale = periodtime in mS
//
//  BIN         NAME        TIME
// 0b00001001  CYC4        0,122 mS
// 0b00011001  CYC8        0,244 mS
// 0b00011001  CYC16       0,488 mS
// 0b00100001  CYC32       0,977 mS
// 0b00101001  CYC64       1,953 mS
// 0b00110001  CYC128      3,906 mS
// 0b00111001  CYC256      7,813 mS
// 0b01000001  CYC512      15,63 mS
// 0b01001001  CYC1024     31,25 mS
// 0b01010001  CYC2048     62,50 mS
// 0b01011001  CYC4096    125,00 mS
// 0b01100001  CYC8192    250,00 mS
// 0b01101001  CYC16384   500,00 mS
// 0b01110001  CYC32768  1000,00 mS
```

**Reply by ASDosjani** [reply]
I just realized that RTC overflow interrupt can't work in power down sleep mode so I guess I'm stuck with PIT and 32s max...

**Reply by hmeijdam** [reply]
Yes, I see that now in the datasheet. In fact now it makes sense to me why there is a PIT and not just the RTC. PIT is primarily intended as the wakeup from powerdown timer I guess.

**Reply by Maaar10** [reply]
> The PIT can only do once every 32 seconds max, but the RTC_CNT interrupt can be set for hours.

This is quite strange to me, because when looking at the datasheet of the ATTINY402, the block diagram suggests that the clock for the PIT goes through the same prescaler as for the RTC. However, putting any value in the prescaler has no effect.
Is this a mistake with the datasheet?

![image](https://github.com/user-attachments/assets/30dc660c-3580-4691-977c-9f7ea2c46ef5)

**Reply by hmeijdam** [reply]
I think where you see 15 bit prescaler in the picture it should say 15 bit prescaler or 15 bit period. If my memory serves me right, the PIT is only looking at the PITCTRLA register, where you set the PIT Period. 

![image](https://github.com/user-attachments/assets/14022b96-b176-4871-9276-c69d76168300)

That acts as a sort of prescaler, which may be the reason the datasheet is written this way.

**Reply by Maaar10** [reply]
I see. That wasnt very obvious to me I guess ^.^'
Thank you for clarifying!

**Comment by ASDosjani** [comment]
Awesome, thank you very much! I will try this out
